File: self-national-census/lib/self/sdk/sdk-go/InMemoryConfigStore.go
package self

import (
	"context"

)

// GetActionIdFunc is a function type for custom action ID generation
type GetActionIdFunc func(ctx context.Context, userIdentifier string, userDefinedData string) (string, error)

// InMemoryConfigStore provides an in-memory implementation of ConfigStore with custom action ID logic
type InMemoryConfigStore struct {
	configs         map[string]VerificationConfig
	getActionIdFunc GetActionIdFunc
}

// Compile-time check to ensure InMemoryConfigStore implements ConfigStore interface
var _ ConfigStore = (*InMemoryConfigStore)(nil)

// NewInMemoryConfigStore creates a new instance of InMemoryConfigStore
func NewInMemoryConfigStore(getActionIdFunc GetActionIdFunc) *InMemoryConfigStore {
	return &InMemoryConfigStore{
		configs:         make(map[string]VerificationConfig),
		getActionIdFunc: getActionIdFunc,
	}
}

// GetActionId uses the custom function to generate action IDs
func (store *InMemoryConfigStore) GetActionId(ctx context.Context, userIdentifier string, userDefinedData string) (string, error) {
	return store.getActionIdFunc(ctx, userIdentifier, userDefinedData)
}

// SetConfig stores a configuration with the given ID
// Returns true if the configuration was newly created, false if it was updated
func (store *InMemoryConfigStore) SetConfig(ctx context.Context, id string, config VerificationConfig) (bool, error) {
	_, existed := store.configs[id]
	store.configs[id] = config
	return !existed, nil
}

// GetConfig retrieves a configuration by ID
	func (store *InMemoryConfigStore) GetConfig(ctx context.Context, id string) (VerificationConfig, error) {
	config, exists := store.configs[id]
	if !exists {
		return VerificationConfig{}, nil
	}
	return config, nil
}

File: self-open-citizen-census/lib/self/sdk/core/src/store/InMemoryConfigStore.ts
import { VerificationConfig } from 'src/types/types.js';
import { IConfigStorage } from './interface.js';

export class InMemoryConfigStore implements IConfigStorage {
  private configs: Map<string, VerificationConfig> = new Map();
  private getActionIdFunc: IConfigStorage['getActionId'];

  constructor(getActionIdFunc: IConfigStorage['getActionId']) {
    this.getActionIdFunc = getActionIdFunc;
  }

  async getActionId(userIdentifier: string, userDefinedData: string): Promise<string> {
    return this.getActionIdFunc(userIdentifier, userDefinedData);
  }

  async setConfig(configId: string, config: VerificationConfig): Promise<boolean> {
    const existed = this.configs.has(configId);
    this.configs.set(configId, config);
    return !existed;
  }

  async getConfig(configId: string): Promise<VerificationConfig> {
    return this.configs.get(configId);
  }
}
